internal/tempfile/tempfile_linux.go

```diff
@@ -0,0 +1,42 @@
+package tempfile
+
+import (
+	"errors"
+	"os"
+	"sync/atomic"
+	"syscall"
+
+	"golang.org/x/sys/unix"
+)
+
+var unsupportedTmpFile = new(int32) //nolint:gochecknoglobals
+
+// Create creates a temporary file that will be automatically deleted on close.
+func Create(dir string) (*os.File, error) {
+	if atomic.LoadInt32(unsupportedTmpFile) == 1 {
+		// already tried O_TMPFILE, was unsupported, fall back to generic
+		// Unix method.
+		return createUnixFallback(dir)
+	}
+
+	// on reasonably modern Linux (3.11 and above) O_TMPFILE is supported,
+	// which creates invisible, unlinked file in a given directory.
+
+	fd, err := unix.Open(dir, unix.O_RDWR|unix.O_TMPFILE|unix.O_CLOEXEC, permissions)
+	if err == nil {
+		return os.NewFile(uintptr(fd), ""), nil
+	}
+
+	if errors.Is(err, syscall.EISDIR) || errors.Is(err, syscall.EOPNOTSUPP) {
+		// O_TMPFILE is unsupported, fall back and prevent future attempts.
+		atomic.StoreInt32(unsupportedTmpFile, 1)
```

Comment:
Preventing future attempts with a global is probably not a good idea:

Whether `O_TMPFILE` is supported does not only depend on the kernel version, but also on the file system and thus mount that this is tried on.

For example, [CephFS](https://tracker.ceph.com/issues/18477) and [SSHFS](https://sourceforge.net/p/fuse/mailman/message/33646980/) do not support `O_TMPFILE` currently.

You could easily have a run of Kopia where tepmfiles are created on a non-supporting and a supporting system, and then it wouldn't work on the supporting one, even though it could.

(I'm aware that for `kopia-map` files this is unlikely, but now that this is a general-purpose library, other future uses could appear.)

So this global should be _per mount_.

But from my experience, there is no significant drawback in always trying `O_TMPFILE` first, and falling back on failure, without remembering the result, for multiple reasons:

* Most file systems support `O_TMPFILE`.
* For those that don't, returning `EOPNOTSUPP (Operation not supported)` should be very fast, as it is a decision that can be made early in the syscall.